packages/@sanity/form-builder/src/inputs/ObjectInput/Field.tsx

```diff
@@ -1,101 +1,93 @@
-      field,
-      level,
-      onFocus,
-      onBlur,
-      markers,
-      focusPath,
-      filterField,
-      compareValue,
-      presence,
-    } = this.props
-    if (typeof value !== 'undefined') {
-      const expectedType = field.type.name
-      const actualType = resolveTypeName(value)
-      // todo: we should consider removing this, and not allow aliasing native types
-      // + ensure custom object types always gets annotated with _type
-      const isCompatible = actualType === field.type.jsonType
-      if (expectedType !== actualType && !isCompatible) {
-        return (
-          <div className={styles.root}>
-            <Fieldset legend={field.type.title} level={level} presence={presence}>
-              <InvalidValue
-                value={value}
-                onChange={this.handleChange}
-                validTypes={[field.type.name]}
-                actualType={actualType}
-                ref={this.setInput}
-              />
-            </Fieldset>
-          </div>
-        )
+export const Field = forwardRef(function Field(props: FieldProps, forwardedRef: ForwardedRef<any>) {
+  const {
+    value,
+    readOnly,
+    field,
+    level,
+    onChange,
+    onFocus,
+    onBlur,
+    markers,
+    focusPath,
+    filterField,
+    compareValue,
+    presence,
+  } = props
+
+  const handleChange = useCallback(
+    (event) => {
+      if (!field.type.readOnly) {
+        onChange(event, field)
       }
+    },
+    [onChange, field]
+  )
+
+  if (typeof value !== 'undefined') {
+    const expectedType = field.type.name
+    const actualType = resolveTypeName(value)
+    // todo: we should consider removing this, and not allow aliasing native types
+    // + ensure custom object types always gets annotated with _type
+    const isCompatible = actualType === field.type.jsonType
+    if (expectedType !== actualType && !isCompatible) {
+      return (
+        <FormFieldSet title={field.type.title} level={level} __unstable_presence={presence}>
+          <InvalidValueInput
+            value={value}
+            onChange={handleChange}
+            validTypes={[field.type.name]}
+            actualType={actualType}
+            ref={forwardedRef}
+          />
+        </FormFieldSet>
+      )
     }
-    return (
-      <div className={styles.root}>
-        <FormBuilderInput
-          value={value}
-          type={field.type}
-          onChange={this.handleChange}
-          path={[field.name]}
-          onFocus={onFocus}
-          onBlur={onBlur}
-          readOnly={readOnly || field.type.readOnly}
-          focusPath={focusPath}
-          filterField={filterField}
-          markers={markers}
-          compareValue={compareValue}
-          level={level}
-          presence={presence}
-          ref={this.setInput}
-        />
-      </div>
-    )
   }
-}
+  return (
+    <FormBuilderInput
+      value={value}
+      type={field.type}
+      onChange={handleChange}
+      path={[field.name]}
```

Comment:
Should this `path` value be memoized?

Reply:
Fixed in https://github.com/sanity-io/sanity/commit/b2f91754eac79fb9ef693cb362d4af0148568a25